ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/core/tests/internal/PDEXmlProcessorFactoryTest.java

```diff
@@ -0,0 +1,344 @@
+		DocumentBuilder documentBuilder = PDEXmlProcessorFactory.createDocumentBuilderWithErrorOnDOCTYPE();
+		testParseXmlWithExternalEntity(documentBuilder, this::createNormalXml);
+	}
+
+	public void testParseXmlWithExternalEntity(DocumentBuilder builder, IntFunction<InputStream> xmlSupplier)
+			throws Exception {
+		try (Server httpServerThread = new Server()) {
+			Document document;
+
+			try (InputStream xmlStream = xmlSupplier.apply(httpServerThread.getLocalPort())) {
+				document = builder.parse(xmlStream);
+			}
+			Element root = document.getDocumentElement();
+
+			assertEquals("Body", root.getTagName());
+			String value = root.getChildNodes().item(0).getNodeValue();
+			assertFalse("Parser injected secret: " + value, value.contains("secret"));
+		} finally {
+			cleanup();
+		}
+	}
+
+	@Test
+	public void testTransformXmlWithExternalEntity() throws Exception {
+		TransformerFactory transformerFactory = PDEXmlProcessorFactory.createTransformerFactoryWithErrorOnDOCTYPE();
+		try {
+			testParseXmlWithExternalEntity(transformerFactory, this::createMalciousXml);
+			assertTrue("TransformerException expected", false);
+		} catch (TransformerException e) {
+			String message = e.getMessage();
+			assertTrue(message, message.contains("DTD"));
+		}
+	}
+
+	@Test
+	public void testTransformXmlWithoutExternalEntity() throws Exception {
+		TransformerFactory transformerFactory = PDEXmlProcessorFactory.createTransformerFactoryWithErrorOnDOCTYPE();
+		testParseXmlWithExternalEntity(transformerFactory, this::createNormalXml);
+	}
+
+	public void testParseXmlWithExternalEntity(TransformerFactory transformerFactory,
+			IntFunction<InputStream> xmlSupplier) throws Exception {
+		try (Server httpServerThread = new Server()) {
+			String formatted;
+
+			try (InputStream xmlStream = xmlSupplier.apply(httpServerThread.getLocalPort())) {
+				Transformer xformer = transformerFactory.newTransformer();
+				Source source = new StreamSource(xmlStream);
+				try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+					Result result = new StreamResult(outputStream);
+					xformer.transform(source, result);
+					formatted = outputStream.toString(StandardCharsets.UTF_8);
+				}
+			}
+			assertTrue(formatted, formatted.contains("<Body>"));
+			assertFalse("Formatter injected secret: " + formatted, formatted.contains("secret"));
+		} finally {
+			cleanup();
+		}
+	}
+
+	private InputStream createMalciousXml(int localPort) {
+		try {
+			Path tempSecret = Files.createTempFile("test", ".txt");
+			tmpFiles.add(tempSecret);
+			Files.writeString(tempSecret, "secret");
+			Path tempDtd = Files.createTempFile("test", ".dtd");
+			tmpFiles.add(tempDtd);
+			String dtdContent = "<!ENTITY % var1 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY var4 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY % var2 \"<!ENTITY var3 SYSTEM 'http://localhost:" + localPort + "/?%var1;'>\">\n" //
+					+ "%var2;\n";
```

Comment:
```suggestion
			String dtdContent = String.format("""
					<!ENTITY % var1 SYSTEM "%s">
					<!ENTITY var4 SYSTEM "%s">
					<!ENTITY % var2 "<!ENTITY var3 SYSTEM 'http://localhost:%s/?%var1;'>">
					%var2;
					""", tempSecret.toUri().toURL(), tempSecret.toUri().toURL(), localPort);
```
A text-block and String.format would make this simpler to read. Once String-Templates are available in a future Java version and we can use that version, it can be simply replaced to use a String-template.

Reply:
it's not obvious to me as a reader, that the "%"  does not need to be escaped and harder to understand which parameter lands where. I prefer the old-school thing.

Reply:
ACK.